nixos/modules/services/misc/matrix-appservices/as-formats.nix

```diff
@@ -0,0 +1,160 @@
+{ name, systemConfig, asConfig, lib, pkgs, ... }:
+
+with lib;
+let
+  inherit (systemConfig.services.matrix-appservices)
+    homeserverURL
+    homeserverDomain;
+  package = asConfig.package;
+  pname = getName package;
+  command = "${package}/bin/${pname}";
+
+  mautrix = {
+    startupScript = ''
+      ${command} --config=$SETTINGS_FILE \
+        --registration=$REGISTRATION_FILE
+    '';
+
+    settings = {
+      homeserver = {
+        address = homeserverURL;
+        domain = homeserverDomain;
+      };
+
+      appservice = with asConfig; {
+        address = "http://${host}:${toString port}";
+
+        hostname = host;
+        inherit port;
+
+        state_store_path = "$DIR/mx-state.json";
+        # mautrix stores the registration tokens in the config file
+        as_token = "$AS_TOKEN";
+        hs_token = "$HS_TOKEN";
+      };
+
+      bridge = {
+        username_template = "${name}_{userid}";
+        permissions = {
+          ${homeserverDomain} = "user";
+        };
+      };
+    };
+  };
+
+in
+{
+  other = {
+    description = ''
+      No defaults will be set.
+    '';
+  };
+
+  matrix-appservice = {
+    startupScript = ''
+      ${command} \
+        --config=$SETTINGS_FILE \
+        --port=$(echo ${asConfig.listenAddress} | sed 's/.*://') \
```

Comment:
Where is asConfig.listenAddress defined?

Reply:
ohh my bad, its not used, I was debating between using `listenAddress` or `host`/`port` options and decided on using the latter so I could set a default host. I actually think I'm just going to remove the `matrix-appservice` format, since a user would have to override most of it anyways. And I'm not currently using and `matrix-appservice` bridges, so I can't test it.